### DISCUSSION TITLE
how can I export or import data in Uint8Array format when using wasm version?

### DISCUSSION THREAD

**Discussion by liushzhql** [isaac-mason/recast-navigation-js#203] [Q&A] [answered]
When using the WebAssembly (wasm) version, how can I export or import navMesh and TileCache data in Uint8Array format? The code file being used is .packages/recast-navigation-wasm/dist/recast-navigation.wasm-compat.js.

**Comment by isaac-mason** [comment]
Hi, `@recast-navigation/wasm` isn't really meant for direct use. It's used internally by the main package (`recast-navigation` / `@recast-navigation/core`). There isn't a "wasm" and "non-wasm" version.

You can find instructions for using `importNavMesh` and `exportNavMesh` methods in the README: https://github.com/isaac-mason/recast-navigation-js#exporting-and-importing

There's also a storybook you can reference: https://github.com/isaac-mason/recast-navigation-js/blob/main/packages/recast-navigation/.storybook/stories/importing-and-exporting.stories.tsx

If you're interested in how the internal `@recast-navigation/wasm` package works for whatever reason, you can find autogenerated docs here:
- https://docs.recast-navigation-js.isaacmason.com/classes/index.Wasm.NavMeshImporter.html
- https://docs.recast-navigation-js.isaacmason.com/classes/index.Wasm.NavMeshExporter.html

**Reply by liushzhql** [reply]
Thank you for your reply. I have read the relevant source code and gained a further understanding of the workflow of this project. However, the way WebAssembly is initialized in the project does not work in my target environment, so I have temporarily given it up. Thank you again anyway.

**Reply by isaac-mason** [reply]
No worries at all!

> However, the way WebAssembly is initialized in the project does not work in my target environment, so I have temporarily given it up.

Could you share more details about this? If there's an improvement that can be made, I'd be keen to learn more 🙂

**Reply by liushzhql** [reply]
I'm currently facing a new challenge: how to make the package `packages/recast-navigation-core` compatible with the asmjs runtime? Inside, the reference to the `@recast-navigation/wasm` package is actually using the file `recast-navigation.wasm-compat.js`. I've already modified the compilation configuration of `@recast-navigation/wasm` to generate the file `./dist/recast-navigation.asm.js`. However, I'm unable to correctly import it in `packages/recast-navigation-core/src/raw-module.ts`. Neither importing the existing `@recast-navigation/wasm/wasm` nor the newly added `@recast-navigation/asm` works. Below is a snippet of the modified `package.json` from the `@recast-navigation/wasm` package:
```
  "type": "module",
  "main": "./dist/recast-navigation.wasm-compat.js",
  "types": "./dist/recast-navigation.d.ts",
  "exports": {
    ".": {
      "import": "./dist/recast-navigation.wasm-compat.js"
    },
    "./wasm": {
      "import": "./dist/recast-navigation.wasm.js"
    },
    "./wasm-compat": {
      "import": "./dist/recast-navigation.wasm-compat.js"
    },
    "./asm": {
      "import": "./dist/recast-navigation.asm.js"
    }
  },
  "files": [
    "dist/recast-navigation.d.ts",
    "dist/recast-navigation.wasm-compat.js",
    "dist/recast-navigation.wasm.js",
    "dist/recast-navigation.wasm.wasm",
    "dist/recast-navigation.asm.js",
    "README.md",
    "LICENSE"
  ],
```

**Reply by isaac-mason** [reply, marked as answer]
There's a related issue for letting users choose their wasm build, I'm not working on it right now though: https://github.com/isaac-mason/recast-navigation-js/issues/164

I'm hoping in future I can support something like this in user-land:

```ts
import { init } from 'recast-navigation'
import MyChosenRecastBuild from '@recast-navigation/wasm/wasm-compat'

await init(MyChosenRecastBuild)
```

Also, I haven't tested what the performance is like with asmjs, but I suspect it will be very bad 🙂 just a warning.

---

Regarding your immediate issue, there's a few things we need to do to fix those entrypoints:
- those entrypoints also need a `"types"` entry for typescript support
- consuming packages (including e.g. `@recast-navigation/core`) need to use a `tsconfig.json` `moduleResolution` option compatible with entrypoints. Right now the packages use `"Node"`, but would need to use an option like `"Bundler"`

As a work-around for testing, you could change the `main` export from `./dist/recast-navigation.wasm-compat.js` to `./dist/recast-navigation.asm.js`

**Reply by liushzhql** [reply]
Thank you for your response and reminder. I am considering using `asm.js` primarily for backward compatibility with platforms that do not support `wasm`. Therefore, I hope for a mechanism that allows for runtime environment detection and dynamic importing.

I look forward to future updates that will support this. Thank you once again.
